maltoolbox/attackgraph/attackgraph.py

```diff
@@ -715,31 +695,54 @@ def regenerate_graph(self) -> None:
 
     def add_node(
             self,
-            node: AttackGraphNode,
-            node_id: Optional[int] = None
-        ) -> None:
-        """Add a node to the graph
+            lg_attack_step: LanguageGraphAttackStep,
+            node_id: Optional[int] = None,
+            model_asset: Optional[ModelAsset] = None,
+        ) -> AttackGraphNode:
+        """Create and add a node to the graph
         Arguments:
-        node    - the node to add
-        node_id - the id to assign to this node, usually used when loading
-                  an attack graph from a file
+        lg_attack_step  - the language graph attack step that corresponds to
+                          the attack graph node to create
+        node_id         - id to assign to the newly created node, usually
+                          provided only when loading an existing attack graph
+                          from a file. If not provided the id will be set to
+                          the next highest id available.
+        model_asset     - the model asset that corresponds to the attack step
+                          node. While optional it is highly recommended that
+                          this be provided. It should only be ommitted if the
+                          model which was used to generate the attack graph is
+                          not available when loading an attack graph from a
+                          file.
+
+        Return:
+        The newly created attack step node.
         """
+        node_id = node_id if node_id is not None else self.next_node_id
+        if node_id in self._id_to_node:
+            raise ValueError(f'Node index {node_id} already in use.')
+        self.next_node_id = max(node_id + 1, self.next_node_id)
+
         if logger.isEnabledFor(logging.DEBUG):
             # Avoid running json.dumps when not in debug
-            logger.debug(f'Add node \"{node.full_name}\" '
-                f'with id:{node_id}:\n' \
-                + json.dumps(node.to_dict(), indent = 2))
+            logger.debug('Create and add to attackgraph node of type "%s" '
+                'with id:%d.\n' % (
+                    lg_attack_step.full_name,
+                    node_id
+                ))
 
-        if node.id in self._id_to_node:
-            raise ValueError(f'Node index {node_id} already in use.')
 
-        node.id = node_id if node_id is not None else self.next_node_id
-        self.next_node_id = max(node.id + 1, self.next_node_id)
+        node = AttackGraphNode(
+            node_id = node_id,
+            lg_attack_step = lg_attack_step,
+            model_asset = model_asset
+        )
 
         self.nodes.append(node)
```

Comment:
BTW, can't we make the node list in the attack graph be a set? Will this break anything? Conceptually, nodes in a graph are unordered. Also, a node should only appear once given that each node has a compromised_by list. This means that using a list which keeps the order of items and allows for multiplicity is not needed.

Reply:
Better still, we already had id keyed dictionaries both both the nodes and the attackers.